sdk/include/platform/sunburst/platform-gpio.hh

```diff
@@ -1,78 +1,186 @@
 #pragma once
 #include <cdefs.h>
 #include <stdint.h>
+#include <utils.hh>
 
 /**
- * Represents the state of the Sonata's joystick.
- *
- * Note, that up to three of the bits may be asserted at any given time.
- * There may be up to two cardinal directions asserted when the joystick is
- * pushed in a diagonal and the joystick may be pressed while being pushed in a
- * given direction.
- */
-enum class SonataJoystick : uint8_t
-{
-	Left    = 1 << 0,
-	Up      = 1 << 1,
-	Pressed = 1 << 2,
-	Down    = 1 << 3,
-	Right   = 1 << 4,
-};
-
-/**
- * A Simple Driver for the Sonata's GPIO.
+ * A simple driver for the Sonata's GPIO. This struct represents a single
+ * GPIO instance, and the methods available to interact with that GPIO.
  *
  * Documentation source can be found at:
- * https://github.com/lowRISC/sonata-system/blob/1a59633d2515d4fe186a07d53e49ff95c18d9bbf/doc/ip/gpio.md
+ * https://github.com/lowRISC/sonata-system/blob/9f794fe3bd4eec8d1a01ee81da97a7f2cec0b452/doc/ip/gpio.md
  *
  * Rendered documentation is served from:
  * https://lowrisc.org/sonata-system/doc/ip/gpio.html
  */
-struct SonataGPIO
+template<
+  /**
+   * The mask of bits of the `output` register that contain meaningful
+   * GPIO output.
+   */
+  uint32_t OutputMask = 0xFFFF'FFFF,
+  /**
+   * The mask of bits of the `input` and `debouncedInput` registers that
+   * contain meaningful GPIO input.
+   */
+  uint32_t InputMask = OutputMask,
```

Comment:
Should this be negated?

Reply:
I _think_ this is fine: most GPIO drivers can be configured to either I or O, these days.

Reply:
Yeah I was confused.